### DISCUSSION TITLE
Prevent Mac sleep during backup

### DISCUSSION THREAD

**Discussion by stefanbuehler** [borgbase/vorta#845] [Ideas]
Mac falls asleep during backup, resulting in low level errors.

It would be great to have a recommended way to prevent this. Ideally a setting in Vorta.

I have also considered whether this could be achieved by putting caffeinate in the shell commands to run before backup, but from my understanding this would not work, since caffeinate never terminates, it just keeps the Mac from sleeping while it runs.

Of course sleep can be prevented altogether which fixes the problem but is not very environment friendly.

**Comment by samuel-w** [comment]
There is a caffeine python module: https://pypi.org/project/caffeine/
And can use systemd-inhibit for Linux: https://github.com/systemd/python-systemd. Most Linux distros use systemd so we should be fine with just that.
Just have to put it at the start of init and the end of finished_event in borg_thread

**Comment by stefanbuehler** [comment]
Yes, but would it be ok to simply make this the default behaviour (that sleep is not allowed while the borg command runs)? For me personally this would be perfectly fine. If no configuration is required, this appears to be relatively easy to implement.

**Comment by m3nu** [comment]
This would need a OS-independent base class and then OS-specific implementations again.

We also can't add the `caffeine` package as dependency, since it's too old and unmaintained. But reusing the code may be possible if the copyright can be clarified. https://github.com/jpn--/caffeine/blob/master/caffeine.py

**Comment by stefanbuehler** [comment]
I have in the meantime also explored if the problem can be solved by using pmset as pre-backup and post-backup shell commands to disable sleep and then reenable. However, the problem with that is that it would have to be run as root.

**Comment by samuel-w** [comment]
It's MIT licensed https://github.com/jpn--/caffeine/blob/master/LICENSE.txt, which is GPL3 compatable

**Comment by stefanbuehler** [comment]
My preliminary fix now is to run "caffeinate&" as pre-borg shell command, and "killall caffeinate" as post-borg shell command. A bit dirty, since it messes with other caffeinate commands in case there are any. But seems to work. The built-in solution discussed above would still be better, of course.

**Comment by unilynx** [comment]
> We also can't add the `caffeine` package as dependency, since it's too old and unmaintained. But reusing the code may be possible if the copyright can be clarified. https://github.com/jpn--/caffeine/blob/master/caffeine.py

https://github.com/jpn--/caffeine/blob/master/LICENSE.txt shows a MIT license, and the name in the license file matches the author of the repository

the code itself traces its origin back to another repository, which also indicates it should be considered MIT licensed - https://github.com/pudquick/pypmset

so I think you would be in the clear to copy this code.

**Comment by cd-a** [comment]
Is there been any way to fix this properly yet?

On my Mac, sometimes the backups run during the night, sometimes not, and sometimes it just hangs in the morning, never recovering until I manually stop and restart the backup.

**Comment by real-yfprojects** [comment]
I found these packages with a quick search:

https://github.com/np-8/wakepy

**Reply by fohrloop** [reply]
Wakepy author here. The newest current version of wakepy (0.9.1) should have most of the users of Vorta covered; macOS  + Linux/Gnome + Linux/KDE (+Windows). For users on for example Arch with a different DE, one could set the [`on_fail` action](https://wakepy.readthedocs.io/v0.9.1/user-guide.html#on-fail-actions) either to "warn" or use custom callable to inform the user and prevent raising the ActivationError when the DE is not supported.

**Comment by rzeigler** [comment]
Cross-linking PR to here: https://github.com/borgbase/vorta/pull/2410
